pkg/storaged/content-views.jsx

```diff
@@ -862,12 +925,126 @@
+        const block = client.blocks[pvol.path];
+        return { type: 'block', block, size: pvol.FreeSize, desc: "", pvol };
+    });
+
+    function validate_pvs(val, vals) {
+        const { layout, pvs } = vals;
+
+        if (layout == "raid0") {
+            if (pvs.length < 2)
+                return _("At least two physical volumes must be selected.");
+        } else if (layout == "raid1") {
+            if (pvs.length < 2)
+                return _("At least two physical volumes must be selected.");
+        } else if (layout == "raid10") {
+            if (pvs.length < 4)
+                return _("At least four physical volumes must be selected.");
+            if (pvs.length % 2 != 0)
+                return _("An even number of physical volumes must be selected.");
+        } else if (layout == "raid4" || layout == "raid5") {
+            if (pvs.length < 3)
+                return _("At least three physical volumes must be selected.");
+        } else if (layout == "raid6") {
+            if (pvs.length < 5)
+                return _("At least five physical volumes must be selected.");
+        } else {
+            if (pvs.length < 1)
+                return _("At least one physical volume must be selected.");
+        }
+    }
+
+    function enough_pvs(vals) {
+        return vals.layout == "linear" || !validate_pvs(null, vals);
+    }
+
+    /* For layouts with redundancy, CreatePlainVolumeWithLayout will
+     * create as many subvolumes as there are selected PVs.  This has
+     * the nice effect of making the calculation of the maximum size of
+     * such a volume trivial.
+     */
+
+    function max_size(vals) {
+        const layout = vals.layout;
+        const pvs = vals.pvs.map(s => s.pvol);
+        const n_pvs = pvs.length;
+        const sum = pvs.reduce((sum, pv) => sum + pv.FreeSize, 0);
+        const min = Math.min.apply(null, pvs.map(pv => pv.FreeSize));
+
+        function metasize(datasize) {
+            const default_regionsize = 2 * 1024 * 1024;
+            const regions = Math.ceil(datasize / default_regionsize);
+            const bytes = 2 * 4096 + Math.ceil(regions / 8);
+            return vgroup.ExtentSize * Math.ceil(bytes / vgroup.ExtentSize);
+        }
+
+        if (layout == "linear") {
+            return vgroup.FreeSize;
+        } else if (layout == "linear_with_pvs") {
+            return sum;
+        } else if (layout == "raid0" && n_pvs >= 2) {
+            return n_pvs * min;
+        } else if (layout == "raid1" && n_pvs >= 2) {
+            return min - metasize(min);
+        } else if (layout == "raid10" && n_pvs >= 4) {
+            return (n_pvs / 2) * (min - metasize(min));
+        } else if ((layout == "raid4" || layout == "raid5") && n_pvs >= 3) {
+            return (n_pvs - 1) * (min - metasize(min));
+        } else if (layout == "raid6" && n_pvs >= 5) {
+            return (n_pvs - 2) * (min - metasize(min));
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17226-20230831-122253-decca6bc-fedora-39-storage//Coverage/lcov/github-pr.diff.gcov.html)